url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Req.java

```diff
@@ -0,0 +1,200 @@
+/*
+    Copyright (c) 2009-2011 250bpm s.r.o.
+    Copyright (c) 2007-2009 iMatix Corporation
+    Copyright (c) 2011 VMware, Inc.
+    Copyright (c) 2007-2011 Other contributors as noted in the AUTHORS file
+
+    This file is part of 0MQ.
+
+    0MQ is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    0MQ is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package zmq;
+
+public class Req extends Dealer {
+
+    
+    //  If true, request was already sent and reply wasn't received yet or
+    //  was raceived partially.
+    private boolean receiving_reply;
+
+    //  If true, we are starting to send/recv a message. The first part
+    //  of the message must be empty message part (backtrace stack bottom).
+    private boolean message_begins;
+    
+    
+    public Req(Ctx parent_, int tid_, int sid_) {
+        super(parent_, tid_, sid_);
+        
+        receiving_reply = false;
+        message_begins = true;
+        options.type = ZMQ.ZMQ_REQ;
+    }
+    
+    @Override
+    public boolean xsend (Msg msg_, int flags_)
+    {
+        //  If we've sent a request and we still haven't got the reply,
+        //  we can't send another request.
+        if (receiving_reply) {
+            throw new IllegalStateException("Cannot send another request");
+        }
+
+        //  First part of the request is the request identity.
+        if (message_begins) {
+            Msg bottom = new Msg();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *bottom*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 59 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).